lib/src/linting/rules/has-discriminator.ts

```diff
@@ -0,0 +1,95 @@
+import assertNever from "assert-never";
+import { compact, flatten } from "lodash";
+import { Locatable } from "../../models/locatable";
+import { EndpointNode } from "../../models/nodes";
+import { DataType, TypeKind, UnionType } from "../../models/types";
+import { inferDiscriminator } from "../../utilities/infer-discriminator";
+import { LintingRule } from "../rule";
+
+/**
+ * Checks that all union types have a discriminator.
+ */
+export const hasDiscriminator: LintingRule = contract => {
+  const topLevelTypes = flatten([
+    ...contract.types,
+    ...contract.endpoints.map(extractEndpointTypes)
+  ]);
+  const unionTypes = flatten(
```

Comment:
is the flatten guaranteed to be only 1 depth?

Reply:
https://lodash.com/docs/4.17.11#flatten yep